src/transformers/adapters/model_mixin.py

```diff
@@ -535,42 +593,13 @@ def pre_transformer_forward(self, **kwargs):
 
 
 @inherit_doc
-class ModelWithHeadsAdaptersMixin(ModelAdaptersMixin):
-    """Mixin adding support for loading/ saving adapters to transformer models with head(s)."""
+class ModelWithHeadsAndNoBaseAdaptersMixin(ModelAdaptersMixin):
+    """Mixin adding support for loading/ saving adapters to transformer models with head(s), and no base model."""
```

Comment:
There's just a nasty infinite recursion that's created by functions patterned like this in ModelWithHeadsAdaptersMixin where the `base_model==model` is true, as you can see below:

```
    def add_adapter(self, adapter_name: str, config=None, overwrite_ok: bool = False):
        self.base_model.add_adapter(adapter_name, config, overwrite_ok=overwrite_ok)
```


There are definitely alternative workarounds where I could explicitly override all of these broken functions. It felt easier for development purposes to create an alternate class `ModelWithHeadsAdaptersMixin` that didn't come with these infinite recursion functions, but had all the head saving / loading mechanisms I wanted.

Would you like me to simply override all of the infinite recursion functions by creating a `T5ModelWithHeadsAdaptersMixin` that extends `ModelWithHeadsAdaptersMixin`? That would also work, and not restructure / split your main adapter classes. The only thing is it will break / need extension if a new infinite recursion function is added to `ModelWithHeadsAdaptersMixin` later. I'll do any this or any other alternate solutions you'd like, you know your code base best.

Reply:
Ah, I see, I totally didn't realize the infinite recursion when looking over the code. I think both of your solution would work. Maybe as an alternative, a simple check within the methods in question would also suffice, e.g. like:
```python
def _add_adapter(self, adapter_name):
    if self.base_model is self:
        super()._add_adapter(adapter_name)
    else:
        self.base_model._add_adapter(adapter_name)
```
That wouldn't require any changes to the overall class structure.

Reply:
The super class of `ModelWithHeads`, namely `ModelAdaptersMixin`, doesn't implement `_add_adapter` though. so this would just cause a pass? No? 

Doing a T5ModelWithHeadsAdaptersMixin should also avoid changing the class structure.

Reply:
It does actually call the implementation in `T5ModelAdaptersMixin` but as this functionality is dependent on the inheritance order in `T5ForConditionalGeneration`, after thinking about it again, I agree that adding a `T5ModelWithHeadsAdaptersMixin` probably would be the best solution.

Reply:
Looking through the code more, I've reconsidered that your initial suggestion to check for self and pass to the super class is probably better. It does indeed rely on the inheritance order.......but its going to be less complex and more readable code.

Reply:
Thank you for the suggestion with doing the check on base model. It was very fast, and means I don't need to refactor the model structure much. That's done now.

Reply:
> Looking through the code more, I've reconsidered that your initial suggestion to check for self and pass to the super class is probably better. It does indeed rely on the inheritance order.......but its going to be less complex and more readable code.

Ok. Please try to document this mechanism and the dependence on inheritance order somewhere in code using comment(s) to make sure it's still understandable in the future. Thanks! :)